contracts/multisig-prover/src/encoding/mvx.rs

```diff
@@ -0,0 +1,537 @@
+use axelar_wasm_std::operators::Operators;
+use bech32::{self, FromBase32};
+use cosmwasm_std::{HexBinary, Uint256};
+use itertools::Itertools;
+use multisig::key::{PublicKey, Signature};
+use multisig::msg::Signer;
+use multiversx_sc_codec::top_encode_to_vec_u8;
+use sha3::{Digest, Keccak256};
+use multisig::worker_set::WorkerSet;
+
+use crate::types::{CommandBatch, Operator};
+use crate::{error::ContractError};
+
+use super::Data;
+
+pub fn command_params(
+    source_chain: String,
+    source_address: String,
+    destination_address: String,
+    payload_hash: &[u8; 32],
+) -> Result<HexBinary, ContractError> {
+    let map_addr_err = |_| ContractError::InvalidMessage {
+        reason: format!(
+            "destination_address is not a valid Mvx address: {}",
+            destination_address
+        ),
+    };
+
+    let (_, data, _) = bech32::decode(&destination_address).map_err(map_addr_err)?;
+    let addr_vec = Vec::<u8>::from_base32(&data).map_err(map_addr_err)?;
+    let destination_address =
+        <[u8; 32]>::try_from(addr_vec).map_err(|_| ContractError::InvalidMessage {
+            reason: format!(
+                "destination_address is not a valid Mvx address: {}",
+                destination_address
+            ),
+        })?;
+
+    Ok(top_encode_to_vec_u8(&(
+        source_chain,
+        source_address,
+        destination_address,
+        payload_hash,
+    ))
+    .expect("couldn't serialize command as mvx")
+    .into())
+}
+
+pub fn transfer_operatorship_params(worker_set: &WorkerSet) -> Result<HexBinary, ContractError> {
+    let mut operators: Vec<([u8; 32], Uint256)> = worker_set
+        .signers
+        .iter()
+        .map(|signer| {
+            (
+                <[u8; 32]>::try_from(signer.pub_key.as_ref())
+                    .expect("couldn't convert pubkey to ed25519 public key"),
+                signer.weight,
+            )
+        })
+        .collect();
+    // The order doesn't matter currently for MultiversX, but keeping this for consistency
+    operators.sort_by_key(|op| op.0.clone());
+    let (addresses, weights): (Vec<[u8; 32]>, Vec<Vec<u8>>) = operators
+        .iter()
+        .map(|operator| (operator.0, uint256_to_compact_vec(operator.1)))
+        .unzip();
+
+    let threshold = uint256_to_compact_vec(worker_set.threshold);
+
+    Ok(top_encode_to_vec_u8(&(addresses, weights, threshold))
+        .expect("couldn't serialize command as mvx")
+        .into())
+}
+
+pub fn msg_digest(command_batch: &CommandBatch) -> HexBinary {
+    // MultiversX is just mimicking EVM here
+    let unsigned = [
+        "\x19MultiversX Signed Message:\n".as_bytes(), // Keccek256 hash length = 32
+        encode(&command_batch.data).as_slice(),
+    ]
+    .concat();
+
+    Keccak256::digest(unsigned).as_slice().into()
```

Comment:
This is different than evm. Notice that we hash twice for evm:
```
    let msg = Keccak256::digest(encode(&command_batch.data).as_slice());

    // Prefix for standard EVM signed data https://eips.ethereum.org/EIPS/eip-191
    let unsigned = [
        "\x19Ethereum Signed Message:\n32".as_bytes(), // Keccek256 hash length = 32
        msg.as_slice(),
    ]
    .concat();

    Keccak256::digest(unsigned).as_slice().into()
```
Is the difference intentional?

Reply:
It is intentional, yes. I saw that `bcs` encoding does it like this so I decided on this approach.

At first I just had the encoded data being hashed without the prefix since on MultiversX there is no need for this prefix.
I added it since I saw that both `abi` and `bcs` encoding use one, I assume to make sure that signed data from a chain can not be used on another chain (although unlikely).
Is there a good reason to have this? Because we can remove it to make this contract and our MultiversX Gateway contract slightly more gas efficient.

Reply:
@raress96 We plan to release a slightly updated gateway version that does the domain separation between chains in a cleaner and more efficient way. In the meantime, I recommend keeping it the same for consistency